src/NBitcoin/Block.cs

```diff
@@ -66,24 +67,69 @@ public BlockHeader()
             this.SetNull();
         }
 
-        public BlockHeader(string hex)
-            : this(Encoders.Hex.DecodeData(hex))
+        public BlockHeader(byte[] bytes)
         {
+            this.ReadWrite(bytes);
         }
 
-        public BlockHeader(byte[] bytes)
+        public BlockHeader(string hex, Network network)
+            : this(hex, network?.Consensus?.ConsensusFactory ?? throw new ArgumentNullException(nameof(network)))
         {
-            this.ReadWrite(bytes);
+
+        }
+
+        public BlockHeader(string hex, Consensus consensus)
+            : this(hex, consensus?.ConsensusFactory ?? throw new ArgumentNullException(nameof(consensus)))
+        {
+
         }
+        public BlockHeader(byte[] data, Network network)
+            : this(data, network?.Consensus?.ConsensusFactory ?? throw new ArgumentNullException(nameof(network)))
+        {
 
-        public static BlockHeader Parse(string hex)
+        }
+
+        public BlockHeader(byte[] data, Consensus consensus)
+            : this(data, consensus?.ConsensusFactory ?? throw new ArgumentNullException(nameof(consensus)))
         {
-            return new BlockHeader(Encoders.Hex.DecodeData(hex));
+        }
+
+        public BlockHeader(byte[] data, ConsensusFactory consensusFactory)
+        {
+            if (data == null)
+                throw new ArgumentNullException(nameof(data));
+
+            if (consensusFactory == null)
+                throw new ArgumentNullException(nameof(consensusFactory));
+
+
+            BitcoinStream bs = new BitcoinStream(data)
+            {
+                ConsensusFactory = consensusFactory
+            };
+
+            this.ReadWrite(bs);
+        }
+
+        public BlockHeader(string hex, ConsensusFactory consensusFactory)
+        {
+            if (hex == null)
```

Comment:
what if hex is string.Empty?

Reply:
Yeah. I guess better to use `string.IsNullOrEmpty()`

Reply:
yeah ok ofcourse.